### PR TITLE
Reorganize the tutorials

### PR DESCRIPTION
**Pull request by melissawm**
This PR is the first in a series of PRs that will implement a reorganization of the tutorials, initially proposed in https://github.com/jax-ml/jax/issues/24632

In this particular commit I am proposing a few things, previously mentioned in https://github.com/jax-ml/jax/issues/24632#issuecomment-3025322421:

- I've changed the User Guides link in the landing page to point to the tutorials, since the user guides have all been moved. I've also added some intro text to the landing page (tutorials.rst) as a placeholder, as I think this needs some copy review - it may well act as a landing page for the docs for folks coming from the main landing page, and seeing just a ToC is not ideal.
- Some of the suggested changes will require other passes (especially through content). Some of the new documents I added are very clearly just stitched together from existing content, and not necessarily meant to be final copy; the main goal of this is to visualize where the existing (and new) documents will land in the site map.
- Instead of "Resources and Advanced Tutorials" I chose to use "Resources and Advanced Guides" as some of these documents are not tutorials. In fact, I would also suggest reviewing some of the documents with some framework (like [[Diatáxis](https://diataxis.fr/)](https://diataxis.fr/) for example) in mind so that they are clearly scoped. For example, I am having trouble differentiating the goals of the [[Pytrees tutorial](https://melissawm.github.io/jax/pytrees.html)](https://melissawm.github.io/jax/pytrees.html) and [[Working with Pytrees](https://melissawm.github.io/jax/working-with-pytrees.html)](https://melissawm.github.io/jax/working-with-pytrees.html) (the Pytrees tutorial is also not included in the original proposed reorg, as far as I can see).
- I did choose to separate sections in the "Resources and Advanced Guides" toctree so we can again have a high-level view of the pages, but these will likely change in a final version.
- I've converted some of the internal links in "How to think in JAX" into absolute urls because this notebook may be opened in colab and this will not have the full documentation context. Let me know if this is not appriopriate and I can revert.

I chose to do smaller PRs to make it easier to review. This one is the largest, just because it is meant to be a coherent reorganization.

### CODE REVIEW

**Comment by melissawm** [comment]
cc @emilyfertig

**Comment by emilyfertig** [comment]
Does https://melissawm.github.io/jax reflect what's in this PR? Some of it looked different to me but maybe I'm looking at it wrong. 

> * Some of the suggested changes will require other passes (especially through content). Some of the new documents I added are very clearly just stitched together from existing content, and not necessarily meant to be final copy; the main goal of this is to visualize where the existing (and new) documents will land in the site map.

This is fine -- the scope of the original outline was to reorganize the existing content and generate little if any new content (though ideas you have for additions/edits are welcome).

> * Instead of "Resources and Advanced Tutorials" I chose to use "Resources and Advanced Guides" as some of these documents are not tutorials. In fact, I would also suggest reviewing some of the documents with some framework (like [[Diatáxis](https://diataxis.fr/)](https://diataxis.fr/) for example) in mind so that they are clearly scoped. For example, I am having trouble differentiating the goals of the [[Pytrees tutorial](https://melissawm.github.io/jax/pytrees.html)](https://melissawm.github.io/jax/pytrees.html) and [[Working with Pytrees](https://melissawm.github.io/jax/working-with-pytrees.html)](https://melissawm.github.io/jax/working-with-pytrees.html) (the Pytrees tutorial is also not included in the original proposed reorg, as far as I can see).

Looks good, and I like the idea of applying a framework. I think we could merge the Pytrees sections (it looks like there's considerable overlap), or split them into a main tutorial and an advanced one (containing the Developer Information from the current "Pytrees tutorial".) The second idea would be similar to what's proposed in the outline (I'm not sure why I didn't propose more consolidation originally, it must have been an oversight).

> * I did choose to separate sections in the "Resources and Advanced Guides" toctree so we can again have a high-level view of the pages, but these will likely change in a final version.

Sounds good -- I think it'd be good to have some grouping here, maybe in dropdowns, like how they're organized in the toctree.

> * I've converted some of the internal links in "How to think in JAX" into absolute urls because this notebook may be opened in colab and this will not have the full documentation context. Let me know if this is not appriopriate and I can revert.

Do a lot of other notebooks have this issue? If so, maybe we should revert this for now, and think about whether we want to do something more system-wide.

File: docs/debugging.md [hunk 1]
```diff
@@ -200,6 +200,117 @@ This can make {func}`jax.debug.callback` useful for general-purpose debugging.
 
 You can learn more about {func}`jax.debug.callback` and other kinds of JAX callbacks in {ref}`external-callbacks`.
 
+## Debugging runtime values
```
Comment (emilyfertig) [inline docs/debugging.md]: Where was this moved/copied from? I don't see in the diff where it was deleted.

Also, could you please consolidate this a little more? E.g. move the "Do you have exploding gradients" to the top, and delete the duplicated material, such as the `jax.debug.print` example.

**Review by emilyfertig — CHANGES_REQUESTED** [review]

File: docs/debugging.md (hunk 1 above)
Comment (melissawm) [reply docs/debugging.md]: This content comes from the Debugging index page (https://github.com/jax-ml/jax/blob/main/docs/debugging/index.md) that I deleted since I thought it would be easier to have all this content in one page. I just reviewed and made it a bit more streamlined, let me know if this is ok.

**Comment by melissawm** [comment]
I also forgot to ask: should I update the "freshness" comments that mention when this content was last reviewed?

About the links in notebooks: I've reviewed them and most already use this strategy (that is, they use absolute links). There are a small number of autodoc links (like "``:func:`jax.grad` ``"). I am happy to make them all clickable in the notebooks if you think it's preferrable.

I've also just updated the preview link, so https://melissawm.github.io/jax should now reflect this PR entirely. 

Thanks!

File: docs/debugging.md [hunk 2]
```diff
@@ -17,7 +17,14 @@ kernelspec:
 
 <!--* freshness: { reviewed: '2024-05-10' } *-->
 
-This section introduces you to a set of built-in JAX debugging methods — {func}`jax.debug.print`, {func}`jax.debug.breakpoint`, and {func}`jax.debug.callback` — that you can use with various JAX transformations.
+Do you have exploding gradients? Are NaNs making you gnash your teeth? Just want
```
Comment (emilyfertig) [inline docs/debugging.md:20]: Could you mention `jax.experimental.checkify` and the debugging flags here too? Maybe delete 
```
— {func}`jax.debug.print`, {func}`jax.debug.breakpoint`, and
{func}`jax.debug.callback` — 
``` 
in this paragraph (so it doesn't become a whole laundry list) and move a couple sentences on checkify/debugging flags from below up to the following "Summary" paragraph.

**Review by emilyfertig — CHANGES_REQUESTED** [review]
Thank you, this looks good! Could you take a look at the presubmit failure?

Thanks for checking the other notebooks -- let's use absolute links if most others already do. It'd be great if you could convert the autodoc links so they're clickable in the notebooks too. And let's update the freshness dates, thanks for catching that.

Also, is it on your agenda to consolidate the Pytrees tutorials later? (If not, I'll try to find time to do that, or make other plans for getting it done).

File: docs/debugging.md (hunk 2 above)
Comment (melissawm) [reply docs/debugging.md:20]: Made some changes, let me know if this works 🙏🏻

**Comment by melissawm** [comment]
Hi @emilyfertig - I rebased since there were some conflicts.

- Pytrees: I do have plans to consolidate those, would you mind if I do it in a follow-up PR?
- Links in notebooks: in the same spirit, I think to be safe and make sure I am using the correct links, I'd like to do this review after this reorg goes in. That way we won't risk using an old link that will immediately stop working.

Let me know what you think.

File: docs/debugging.md (hunk 2 above)
Comment (emilyfertig) [reply docs/debugging.md:20]: Looks good!

**Comment by emilyfertig** [comment]
That sounds great, thanks!

Could you please fix lint_and_typecheck (looks like a `jupytext --sync` thing), and squash to a single commit, and I'll approve?

**Review by emilyfertig — APPROVED** [review]

**Comment by mattjj** [comment]
Hey @melissawm, thanks for this work!

However, we just rolled it back in https://github.com/jax-ml/jax/commit/0abd69624102b55c5bcc40b065c2982fcd857852 because some folks on the team didn't like some of the changes, and wanted to weigh in. Sorry if this feels like getting mixed messages!

We'll follow up soon.

**Comment by mattjj** [comment]
(Sorry for being slow!)

Can we roll this forward with these changes?
1. put the sharded-computation doc link back into the Tutorials section
2. move the "Parallel computation" section to the top of the "Resources and Advanced Guides" 

More generally, we want to avoid having docs changes push parallelism/scaling stuff into the background, since we want to be foregrounding that stuff as much as possible.

Thanks @melissawm!

**Comment by melissawm** [comment]
Will do, thanks for the follow-up @mattjj !
